### PR TITLE
support enterprise_project_id in elb loadbalancer

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

support enterprise_project_id in elb loadbalancer
in order to support enterprise_project_id, the sdk and client of loadbalancer were updated to v2
All test cases were run to check whether it affects other related resources

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
related #1019 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccLBV2'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccLBV2 -timeout 360m -parallel 4
=== RUN   TestAccLBV2Certificate_basic
=== PAUSE TestAccLBV2Certificate_basic
=== RUN   TestAccLBV2Certificate_client
=== PAUSE TestAccLBV2Certificate_client
=== RUN   TestAccLBV2L7Policy_basic
=== PAUSE TestAccLBV2L7Policy_basic
=== RUN   TestAccLBV2L7Rule_basic
=== PAUSE TestAccLBV2L7Rule_basic
=== RUN   TestAccLBV2Listener_basic
=== PAUSE TestAccLBV2Listener_basic
=== RUN   TestAccLBV2LoadBalancer_basic
=== PAUSE TestAccLBV2LoadBalancer_basic
=== RUN   TestAccLBV2LoadBalancer_secGroup
=== PAUSE TestAccLBV2LoadBalancer_secGroup
=== RUN   TestAccLBV2LoadBalancer_withEpsId
=== PAUSE TestAccLBV2LoadBalancer_withEpsId
=== RUN   TestAccLBV2Member_basic
=== PAUSE TestAccLBV2Member_basic
=== RUN   TestAccLBV2Monitor_basic
=== PAUSE TestAccLBV2Monitor_basic
=== RUN   TestAccLBV2Pool_basic
=== PAUSE TestAccLBV2Pool_basic
=== RUN   TestAccLBV2Whitelist_basic
=== PAUSE TestAccLBV2Whitelist_basic
=== CONT  TestAccLBV2Certificate_basic
=== CONT  TestAccLBV2LoadBalancer_withEpsId
=== CONT  TestAccLBV2Pool_basic
=== CONT  TestAccLBV2Monitor_basic
--- PASS: TestAccLBV2Certificate_basic (31.19s)
=== CONT  TestAccLBV2LoadBalancer_secGroup
--- PASS: TestAccLBV2LoadBalancer_withEpsId (44.07s)
=== CONT  TestAccLBV2LoadBalancer_basic
--- PASS: TestAccLBV2Pool_basic (127.78s)
=== CONT  TestAccLBV2Listener_basic
--- PASS: TestAccLBV2LoadBalancer_basic (83.79s)
=== CONT  TestAccLBV2L7Rule_basic
--- PASS: TestAccLBV2LoadBalancer_secGroup (108.10s)
=== CONT  TestAccLBV2L7Policy_basic
--- PASS: TestAccLBV2Monitor_basic (152.58s)
=== CONT  TestAccLBV2Certificate_client
--- PASS: TestAccLBV2Certificate_client (34.61s)
=== CONT  TestAccLBV2Member_basic
--- PASS: TestAccLBV2L7Policy_basic (98.32s)
=== CONT  TestAccLBV2Whitelist_basic
--- PASS: TestAccLBV2Listener_basic (111.44s)
--- PASS: TestAccLBV2L7Rule_basic (136.62s)
--- PASS: TestAccLBV2Whitelist_basic (90.59s)
--- PASS: TestAccLBV2Member_basic (147.98s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       335.218s
```

### CODE REVIEW

File: huaweicloud/lb_v2_shared.go [hunk 1]
```diff
@@ -92,6 +93,46 @@ func waitForLBV2LoadBalancer(networkingClient *golangsdk.ServiceClient, id strin
 	return nil
 }
 
+// for v2 api
+func waitForLBV2LoadBalancer_v2(networkingClient *golangsdk.ServiceClient, id string, target string, pending []string, timeout time.Duration) error {
+	log.Printf("[DEBUG] Waiting for loadbalancer %s to become %s.", id, target)
```
Comment (Lance52259) [inline huaweicloud/lb_v2_shared.go]: the log content shouldn't contains period.

File: huaweicloud/lb_v2_shared.go [hunk 2]
```diff
@@ -92,6 +93,46 @@ func waitForLBV2LoadBalancer(networkingClient *golangsdk.ServiceClient, id strin
...
+				return fmt.Errorf("Error: loadbalancer %s not found: %s", id, err)
+			}
+		}
+		return fmt.Errorf("Error waiting for loadbalancer %s to become %s: %s", id, target, err)
+	}
+
+	return nil
+}
+
+func resourceLBV2LoadBalancerRefreshFunc_v2(networkingClient *golangsdk.ServiceClient, id string) resource.StateRefreshFunc {
```
Comment (Lance52259) [inline huaweicloud/lb_v2_shared.go]: You can control the maximum number of characters in a single line (120) by wrapping lines or reduce the name description of this function.

File: huaweicloud/lb_v2_shared.go [hunk 3]
```diff
@@ -92,6 +93,46 @@ func waitForLBV2LoadBalancer(networkingClient *golangsdk.ServiceClient, id strin
...
+		lb, err := loadbalancers_v2.Get(networkingClient, id).Extract()
+		if err != nil {
+			return nil, "", err
+		}
+
+		return lb, lb.ProvisioningStatus, nil
+	}
+}
+
 func resourceLBV2LoadBalancerRefreshFunc(networkingClient *golangsdk.ServiceClient, id string) resource.StateRefreshFunc {
```
Comment (Lance52259) [inline huaweicloud/lb_v2_shared.go]: ditto (deal with in next commit ?)

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go [hunk 4]
```diff
@@ -208,7 +220,11 @@ func resourceLoadBalancerV2Read(d *schema.ResourceData, meta interface{}) error
 	}
 
 	// fetch tags
-	if resourceTags, err := tags.Get(elbClient, "loadbalancers", d.Id()).Extract(); err == nil {
+	elbV2Client, err := config.ElbV2Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud elb client: %s", err)
+	}
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: two logs (`ElbV2Client` and `Elb_V2Client`) cannot be easily identified.

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go [hunk 5]
```diff
@@ -220,7 +236,7 @@ func resourceLoadBalancerV2Read(d *schema.ResourceData, meta interface{}) error
 
 func resourceLoadBalancerV2Update(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(*Config)
-	elbClient, err := config.ElbV2Client(GetRegion(d, config))
+	elbClient, err := config.Elb_V2Client(GetRegion(d, config))
 	if err != nil {
 		return fmt.Errorf("Error creating HuaweiCloud elb client: %s", err)
 	}
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: ditto

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go [hunk 6]
```diff
@@ -279,7 +295,11 @@ func resourceLoadBalancerV2Update(d *schema.ResourceData, meta interface{}) erro
 
 	// update tags
 	if d.HasChange("tags") {
-		tagErr := UpdateResourceTags(elbClient, d, "loadbalancers", d.Id())
+		elbV2Client, err := config.ElbV2Client(GetRegion(d, config))
+		if err != nil {
+			return fmt.Errorf("Error creating HuaweiCloud elb client: %s", err)
+		}
```
Comment (Lance52259) [inline huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: ditto

File: huaweicloud/lb_v2_shared.go (hunk 2 above)
Comment (Jason-Zhang9309) [reply huaweicloud/lb_v2_shared.go]: fixed

File: huaweicloud/lb_v2_shared.go (hunk 3 above)
Comment (Jason-Zhang9309) [reply huaweicloud/lb_v2_shared.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go (hunk 4 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer.go:226]: fixed

File: huaweicloud/config.go [hunk 7]
```diff
@@ -618,6 +618,10 @@ func (c *Config) ElbV2Client(region string) (*golangsdk.ServiceClient, error) {
 	return c.NewServiceClient("elbv2", region)
 }
 
+func (c *Config) Elb_V2Client(region string) (*golangsdk.ServiceClient, error) {
```
Comment (ShiChangkuo) [inline huaweicloud/config.go]: how about rename the client as `LoadBalancerClient`?
there are many clients for elb service, sugget to add some comments here

File: huaweicloud/endpoints.go [hunk 8]
```diff
@@ -156,6 +156,10 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		Version:          "v2.0",
 		WithOutProjectID: true,
 	},
+	"elb_v2": {
+		Name:    "elb",
+		Version: "v2",
+	},
```
Comment (ShiChangkuo) [inline huaweicloud/endpoints.go]: should add a acc test in `TestAccServiceEndpoints_Network`

File: huaweicloud/resource_huaweicloud_lb_loadbalancer_test.go [hunk 9]
```diff
@@ -307,3 +329,28 @@ resource "huaweicloud_lb_loadbalancer" "loadbalancer_1" {
...
+  tags = {
+    key   = "value"
+    owner = "terraform"
+  }
+
+  timeouts {
+    create = "5m"
+    update = "5m"
+    delete = "5m"
+  }
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_lb_loadbalancer_test.go]: timeoutes block is unnecessray

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go [hunk 10]
```diff
@@ -156,10 +163,14 @@ func resourceLoadBalancerV2Create(d *schema.ResourceData, meta interface{}) erro
 	}
 
 	//set tags
+	elbV2Client, err := config.ElbV2Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating HuaweiCloud elb v2.0 client: %s", err)
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: suggest put those line in head of the function

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]

File: huaweicloud/config.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/config.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer_test.go (hunk 9 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer_test.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go (hunk 10 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: fixed

File: huaweicloud/lb_v2_shared.go (hunk 1 above)
Comment (Jason-Zhang9309) [reply huaweicloud/lb_v2_shared.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go (hunk 4 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go (hunk 5 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: fixed

File: huaweicloud/resource_huaweicloud_lb_loadbalancer.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_lb_loadbalancer.go]: fixed

**Comment by Jason-Zhang9309** [comment]
new test result:
```make testacc TEST='./huaweicloud' TESTARGS='-run TestAccLBV2LoadBalancer_withEpsId'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccLBV2LoadBalancer_withEpsId -timeout 360m -parallel 4
=== RUN   TestAccLBV2LoadBalancer_withEpsId
=== PAUSE TestAccLBV2LoadBalancer_withEpsId
=== CONT  TestAccLBV2LoadBalancer_withEpsId
--- PASS: TestAccLBV2LoadBalancer_withEpsId (52.94s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       52.985s
```

**Review by ShiChangkuo — APPROVED** [review]

**Pull request merged**